src/hotspot/cpu/aarch64/frame_aarch64.cpp

```diff
@@ -262,9 +263,13 @@ bool frame::safe_for_sender(JavaThread *thread) {
     return false;
   }
 
-  // Will the pc we fetch be non-zero (which we'll find at the oldest frame)
+  // Will the pc we fetch be non-zero (which we'll find at the oldest frame) and readable
 
-  if ( (address) this->fp()[return_addr_offset] == NULL) return false;
+  if (!os::is_readable_pointer(this->fp() + return_addr_offset * sizeof(address))) {
```

Comment:
Address computation looks wrong: `fp()` returns an `intptr_t*`, so `return_addr_offset` gets scaled automatically before the addition. Wouldn't `thread->is_in_usable_stack(fp())` be a better check?

Reply:
> Wouldn't thread->is_in_usable_stack(fp()) be a better check?
No, this would check something different.

> Address computation looks wrong
But tests show that it seems to be correct.

Reply:
I think Martin is right.

Reply:
With the first or the second point?

Reply:
To his second point: You have to check the return address location, not just `fp()`.

Reply:
I stand correct, it also helps. I'll change the PR.